File: ContainerWithMostWater_LC11.java
//Time: O(n) optimal , O(n^2) for brute force
//space: o(1) no extra space

class Solution {
    public int maxArea(int[] height) {
        
        int ans = 0;
        int n = height.length;
        
        //brute force
//         for(int i = 0; i < n - 1; i++){
            
//             for(int j = i+1; j < n; j++){
//                 int min = Math.min(height[i], height[j]);
//                 ans = Math.max(ans, (min * (j - i)));
//             }
//         }
        
        //optimal
        
        int left = 0;
        int right = n - 1;
        
        while (left < right){
            int min = Math.min(height[left], height[right]);
            ans = Math.max(ans, (min * (right - left)));
            if(height[left] < height[right])
                left++;
            else
                right--;
        }n
        
        
        return ans;
        
    }
}

File: containerWithMostWater-LC 11.py
# Time Complexity = O(n), Iterating over all elements once
# Space Complexity = O(1)

class Solution:
    def maxArea(self, height: list[int]) -> int:
        if height == None or len(height) == 0:
            return 0
        
        n = len(height)
        maxArea = 0
        
        # Initialising the index of pointers
        left = 0
        right = n - 1
        
        # Base condition to stop the while loop
        while (left <= right):
            maxArea = max(maxArea, (min(height[left], height[right]) * (right - left)))
            
            # If left pointer element is smaller, we shift the left pointer coz there is a possibility of finding a larger element ahead
            if height[left] < height[right]:
                left += 1
            else: 
                right -= 1
                
        return maxArea